contracts/Enforcer.sol

```diff
@@ -1,134 +1,184 @@
-            block.number,
-            endHash,
+        emit Registered(
+            _taskHash,
+            _solverPathRoot,
             executionDepth,
-            customEnvironmentHash,
-            msg.sender
+            _result
         );
-        bonds[msg.sender] += bondAmount;
-
-        emit Registered(executionId, msg.sender, codeContractAddress, _callData);
     }
 
-    /**
-      * @dev dispute is called by challenger to start a new dispute
-      *     assumed that challenger's execution tree is of the same depth as solver's
-      *     in case challenger's tree is shallower, he should use node with zero hash to make it deeper
-      *     in case challenger's tree is deeper, he should submit only the left subtree with the same depth with solver's
-      */
-    function dispute(address codeContractAddress, bytes memory _callData, bytes32 endHash)
+    function dispute(bytes32 _solverPathRoot, bytes32 _challengerPathRoot, EVMParameters memory _parameters)
         public payable
     {
-        bytes32 executionId = keccak256(abi.encodePacked(codeContractAddress, _callData));
+        bytes32 taskHash = parameterHash(_parameters);
+        bytes32 executionId = keccak256(abi.encodePacked(taskHash, _solverPathRoot));
+        ExecutionResult memory executionResult = executions[executionId];
 
-        Execution storage execution = executions[executionId];
-
-        require(execution.startBlock != 0, "Execution does not existed");
-        // executionDepth round plus 1 for submitProof
-        require(execution.startBlock + challengePeriod > block.number + (execution.executionDepth + 1) * verifier.timeoutDuration(),
-                "Execution is out of challenge period");
         require(msg.value == bondAmount, "Bond amount is required");
-        // Do we want to prohibit early?
-        // require(execution.endHash != endHash);
+        require(executionResult.startBlock != 0, "Execution does not exist");
+        // executionDepth round plus 1 for submitProof
+        require(
+            executionResult.startBlock + challengePeriod > block.number + (executionResult.executionDepth + 1) * verifier.timeoutDuration(),
+            "Execution is out of challenge period"
+        );
 
         bonds[msg.sender] += bondAmount;
 
+        // TODO: Verifier needs to support all EVMParameters
         bytes32 disputeId = verifier.initGame(
             executionId,
-            execution.endHash,
-            endHash,
-            execution.executionDepth,
-            execution.customEnvironmentHash,
+            _solverPathRoot,
+            _challengerPathRoot,
+            executionResult.executionDepth,
+            _parameters.customEnvironmentHash,
+            _parameters.codeHash,
+            _parameters.dataHash,
             // challenger
-            msg.sender,
-            codeContractAddress,
-            _callData
+            msg.sender
         );
 
         emit DisputeInitialised(disputeId, executionId);
```

Comment:
this should emit the the same `Registered` events as the `register()` function.

Reply:
Oh wait. Does that makes sense? Now the client has to check if the `Registered` event was for a dispute. Why do that?